wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala

```diff
@@ -30,6 +37,380 @@ sealed abstract class Wallet extends LockedWallet with UnlockedWalletApi {
+  private def getDummyDLCFundingOutputs(
+      amount: CurrencyUnit): Vector[TransactionOutput] = {
+    val key = ECPublicKey.freshPublicKey
+    val spk = P2WSHWitnessSPKV0(MultiSignatureScriptPubKey(2, Seq(key, key)))
+
+    // funded output and the other party's change
+    Vector(TransactionOutput(amount, spk), TransactionOutput(Satoshis.one, spk))
+  }
+
+  override def createDLCOffer(
+      amount: Bitcoins,
+      oracleInfo: OracleInfo,
+      contractInfoVec: Vector[Sha256DigestBE],
+      feeRateOpt: Option[SatoshisPerVirtualByte],
+      locktime: UInt32,
+      refundLocktime: UInt32): Future[DLCOffer] = {
+    logger.debug("Calculating relevant wallet data for DLC Offer")
+
+    val contractInfo =
+      ContractInfo(
+        (Map.newBuilder ++= contractInfoVec.map(sha => (sha, amount.satoshis)))
+          .result())
+
+    val timeouts = DLCTimeouts(5,
+                               BlockStamp(locktime.toInt),
+                               BlockStamp(refundLocktime.toInt))
+
+    val feeRate = feeRateOpt.getOrElse(SatoshisPerVirtualByte.one)
+
+    val eventId = DLCMessage.calcEventId(oracleInfo, contractInfo, timeouts)
+
+    logger.debug("Checking if offer has already been made")
+    for {
+      account <- getDefaultAccountForType(AddressType.SegWit)
+      dlcDbOpt <- dlcDAO.findByEventId(eventId)
+      offer <- dlcDbOpt match {
+        case Some(dlcDb) =>
+          logger.debug("DLC Offer has already been made, returning offer")
+          listUtxos(dlcDb.fundingInputs).map { spendingInfoDbs =>
+            val fundingInputs =
+              spendingInfoDbs
+                .map { db =>
+                  val info =
+                    db.toUTXOSpendingInfo(account, keyManager, dlcDb.network)
+                  (info.outPoint, info.output)
+                }
+            DLCOffer(
+              dlcDb.contractInfo,
+              oracleInfo,
+              keyManager.getRootXPub, // fixme this will need to be changed according to KeyDerivation.md
+              amount.satoshis,
+              fundingInputs,
+              Bech32Address(dlcDb.changeSPK, dlcDb.network),
+              feeRate,
+              dlcDb.timeouts
+            )
+          }
+        case None =>
+          logger.debug("DLC Offer not found, creating a new one")
+          createNewDLCOffer(account = account,
+                            eventId = eventId,
+                            amount = amount,
+                            oracleInfo = oracleInfo,
+                            contractInfo = contractInfo,
+                            feeRate = feeRate,
+                            timeouts = timeouts)
+      }
+    } yield offer
+  }
+
+  private def createNewDLCOffer(
+      account: AccountDb,
+      eventId: Sha256DigestBE,
+      amount: Bitcoins,
+      oracleInfo: OracleInfo,
+      contractInfo: ContractInfo,
+      feeRate: SatoshisPerVirtualByte,
+      timeouts: DLCTimeouts): Future[DLCOffer] = {
+    val dlcDbFs = for {
```

Comment:
you should re-write this `for` expression as this so you don't end up with the type of `Future[(Future[DlcDb], Vector[UTXOSpendingInfo]))` now, usually it is a bad sign when you have nested `Future`. This re-write gives you a clean `Future[(DlcDb, Vector[UTXOSpendingInfo])]` 

```scala
    val dlcDbFs: Future[(DLCDb, Vector[UTXOSpendingInfo])] = for {
      txBuilder <- fundRawTransactionInternal(
        destinations = getDummyDLCFundingOutputs(amount),
        feeRate = feeRate,
        fromAccount = account,
        keyManagerOpt = Some(keyManager)
      )
      utxos = txBuilder.utxoMap
        .filter(_._2.amount != Satoshis.one)
        .map(utxo => (utxo._1, utxo._2.output))
        .toVector

      changeSPK =
      txBuilder.changeSPK.asInstanceOf[WitnessScriptPubKey]
      network = networkParameters.asInstanceOf[BitcoinNetwork]
      _ = logger.debug("DLC Offer data collected, creating database entry")

      dlcDb = {
        DLCDb(
          eventId = eventId,
          oracleInfo = oracleInfo,
          contractInfo = contractInfo,
          timeouts = timeouts,
          isInitiator = true,
          remoteExtPubKey = ExtPrivateKey
            .freshRootKey(SegWitTestNet3Priv)
            .extPublicKey,
          input = amount,
          remoteInput = Satoshis.zero,
          fundingInputs = utxos.map(_._1),
          remoteFundingInputs = Vector.empty,
          feeRate = feeRate,
          changeSPK = changeSPK,
          remoteChangeSPK = P2WSHWitnessSPKV0(EmptyScriptPubKey),
          cetSigs = CETSignatures(PartialSignature.dummyPartialSig,
            PartialSignature.dummyPartialSig,
            PartialSignature.dummyPartialSig),
          fundingSigs =
            FundingSignatures(Vector(PartialSignature.dummyPartialSig)),
          oracleSig = DummySchnorrDigitalSignature,
          network = network
        )
      }

      createdDlc <- dlcDAO.create(dlcDb)
    } yield {
      (createdDlc, utxos)
    }
```